File: main/src/main/java/com/sedmelluq/discord/lavaplayer/natives/vorbis/VorbisDecoderLibrary.java
package com.sedmelluq.discord.lavaplayer.natives.vorbis;

import com.sedmelluq.discord.lavaplayer.natives.ConnectorNativeLibLoader;

import java.nio.ByteBuffer;

class VorbisDecoderLibrary {
    private VorbisDecoderLibrary() {

    }

    static VorbisDecoderLibrary getInstance() {
        ConnectorNativeLibLoader.loadConnectorLibrary();
        return new VorbisDecoderLibrary();
    }

    native long create();

    native void destroy(long instance);

    native boolean initialise(long instance, ByteBuffer infoBuffer, int infoOffset, int infoLength,
                              ByteBuffer setupBuffer, int setupOffset, int setupLength);

    native int getChannelCount(long instance);

    native int input(long instance, ByteBuffer directBuffer, int offset, int length);

    native int output(long instance, float[][] channels, int length);
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/natives/vorbis/VorbisDecoderLibrary.kt
package com.sedmelluq.discord.lavaplayer.natives.vorbis

import com.sedmelluq.discord.lavaplayer.natives.ConnectorNativeLibLoader
import java.nio.ByteBuffer

internal class VorbisDecoderLibrary private constructor() {
    external fun create(): Long
    external fun destroy(instance: Long)
    external fun initialise(instance: Long, infoBuffer: ByteBuffer?, infoOffset: Int, infoLength: Int, setupBuffer: ByteBuffer?, setupOffset: Int, setupLength: Int): Boolean

    external fun getChannelCount(instance: Long): Int
    external fun input(instance: Long, directBuffer: ByteBuffer?, offset: Int, length: Int): Int
    external fun output(instance: Long, channels: Array<FloatArray?>?, length: Int): Int

    companion object {
        @kotlin.jvm.JvmStatic
        val instance: VorbisDecoderLibrary
            get() {
                ConnectorNativeLibLoader.loadConnectorLibrary()
                return VorbisDecoderLibrary()
            }
    }
}
